src/useObservable.ts

```diff
@@ -67,29 +67,26 @@ export function useObservable<ObservableType extends Observable<any>, InitialVal
 
     cache.set(observable, entry as CacheRecord<ObservedValueOf<ObservableType>>)
   }
+  const instance = cache.get(observable)!
 
-  const store = useMemo(() => {
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    const instance = cache.get(observable)!
-    return {
-      subscribe: (onStoreChange: () => void) => {
-        const subscription = instance.observable.subscribe(onStoreChange)
-        return () => {
-          subscription.unsubscribe()
-        }
-      },
-      getSnapshot: () => {
-        if (instance.error) {
-          throw instance.error
-        }
-        return instance.snapshot
-      },
-    }
-  }, [observable])
+  const subscribe = useCallback(
```

Comment:
The rewrite to `useCallback` on just `subscribe`, instead of keeping the `useMemo`, is to avoid agressive subscribe + unsubscribe loops.

The Compiler looks at what is actually used inside the `useMemo`, which is `instance.observable`, `instance.snapshot` and `instance.error`.
In a sense, it's as if it rewrote our code to be:
```ts

const instance = useMemo(() => cache.get(observable), [observable])
const store = useMemo(() => {
  // ...
}, [instance.observable, instance.snapshot, instance.error])
```
This is causing `store` to change identity every time `snapshot` changes.

So by splitting them up, to use `useCallback` instead, this is what happens instead (conceptually):

```ts
const instance = useMemo(() => cache.get(observable), [observable])
const subscribe = useCallback(() => {
  // ...
}, [instance.observable])
```
And now we avoid the problem, as `instance.observable` is stable. `useSyncExternalStore` doesn't care if `getSnapshot` or `getServerSnapshot` changes identity between renders, it only requires `subscribe` to be memoized correctly.

Reply:
Oh, this is great - thanks for the explainer!